src/main.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/testcontainers/testcontainers-go"
+	tcollama "github.com/testcontainers/testcontainers-go/modules/ollama"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"regexp"
+)
+
+type TagResponse struct {
+	Models []struct {
+		Name string `json:"name"`
+	} `json:"models"`
+}
+
+type GenerateRequest struct {
+	Model  string `json:"model"`
+	Stream bool   `json:"stream"`
+	Prompt string `json:"prompt"`
+}
+
+func startContainer(image string) (string, error) {
+	ctx := context.Background()
+	ollamaContainer, err := tcollama.Run(ctx, image, testcontainers.CustomizeRequestOption(func(req *testcontainers.GenericContainerRequest) error {
+		req.Name = normalizeString(image)
+		req.Reuse = true
+		return nil
+	}))
+	if err != nil {
+		log.Fatalf("failed to start container: %s", err)
+		return "", err
+	}
+	return ollamaContainer.Endpoint(ctx, "http")
+}
+
+func normalizeString(input string) string {
+	// Define a regular expression that matches any character that is not alphanumeric, _, or -
+	re := regexp.MustCompile(`[^a-zA-Z0-9_-]`)
+
+	// Replace all matches with a dash
+	normalized := re.ReplaceAllString(input, "-")
+
+	return normalized
+}
+
+func generate(url string, prompt string) error {
+	tagsResp, err := http.Get(fmt.Sprintf("%s/api/tags", url))
+	if err != nil {
+		return err
+	}
+	defer func(Body io.ReadCloser) {
+		_ = Body.Close()
+	}(tagsResp.Body)
+
+	var tags TagResponse
+	if err := json.NewDecoder(tagsResp.Body).Decode(&tags); err != nil {
+		return err
+	}
+
+	model := tags.Models[0].Name
+	genReq := GenerateRequest{
+		Model:  model,
+		Stream: true,
+		Prompt: prompt,
+	}
+
+	reqBody, err := json.Marshal(genReq)
+	if err != nil {
+		return err
+	}
+
+	resp, err := http.Post(fmt.Sprintf("%s/api/generate", url), "application/json", bytes.NewBuffer(reqBody))
+	if err != nil {
+		return err
+	}
+	defer func(Body io.ReadCloser) {
+		_ = Body.Close()
+	}(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("failed to generate: %s", resp.Status)
+	}
+
+	decoder := json.NewDecoder(resp.Body)
```

Comment:
Any particular reason to not use `json.Unmarshal`? See eg https://gobyexample.com/json

Reply:
Using decode results in less code. To use marshall/unmarshall I have to first convert `resp.Body` to an array of bytes:
`body, err := io.ReadAll(resp.Body)`
I'm also not sure how to stream the response by using unmarshall, is it even possible?